rt/x86-64-linux/System.v3

```diff
@@ -8,7 +8,120 @@
 // XXX: use on-stack storage for small IO buffers
 // XXX: what about GC moving arrays during IO operations?
 // XXX: merge common code with darwin/System.v3 into a common Posix implementation
+
+// XXX: Unsound as we need to enforce that ipv4 and ipv6 are exactly 4 and 16
+// u8s long respectively. One way to achieve this could be to create two more
+// types, Inet4Array and Inet6Array, like so:
+// class Inet4Array {
+// 	def addr: Array<u8> = Array.new(4);
+// }
+// class Inet6Array {
+// 	def addr: Array<u8> = Array.new(16);
+// }
+// type InetAddress {
+// 	case Inet4(addr: Inet4Array);
+// 	case Inet6(addr: Inet6Array);
+// }
+/**
+ * Represents an internet address. It can be one of Inet4 for IPv4 addresses or
+ * Inet6 for IPv6 addresses.
+ */
+type InetAddress {
+	case Inet4(addr: Array<u8>);
+	case Inet6(addr: Array<u8>);
+}
+
+class SocketBase {
+	// IP address of socket
+	def addr: InetAddress;
+	// Associated file descriptor of socket
+	var fd: int;
+	// Local port the socket is connected to
+	def localPort: int;
+	// Server port the socket is connected to
+	def port: int;
+
+	new(addr, localPort, port) {
+		var family: int;
+		match (addr) {
+			Inet4(addrV4) => family = LinuxConst.AF_INET;
+			Inet6(addrV6) => family = LinuxConst.AF_INET6;
+		}
+		fd = System.socket(family, LinuxConst.SOCK_STREAM /* type */, 0 /* protocol */);
```

Comment:
Another thing to think about is auto-closing on finalization of this object. I've started to do that with `RiGc.registerFinalizer`. You have to split the object into a GC'able part and a part (maybe just a closure) that closes the file descriptor.

Reply:
> You have to split the object into a GC'able part and a part (maybe just a closure) that closes the file descriptor.

Not sure I follow. Would `RiGc.registerFinalizer(sock, sock.close())` not be sufficient? Would you have an example for the same?